efi/preinstall/check_tcglog.go

```diff
@@ -0,0 +1,610 @@
+		}
+	}
+
+	// Read the actual PCR values from the TPM.
+	var pcrs []int
+	for _, pcr := range supportedPcrs {
+		pcrs = append(pcrs, int(pcr))
+	}
+	selections := tpm2.PCRSelectionList{{Hash: alg, Select: pcrs}}
+	_, values, err := tpm.PCRRead(selections)
+	if err != nil {
+		return nil, err
+	}
+
+	// Record these values on the results
+	if err := results.setPcrValues(values); err != nil {
+		return nil, fmt.Errorf("cannot process PCR values from TPM: %w", err)
+	}
+
+	return results, nil
+}
+
+// tcglogPhase describes the phase of a TCG log
+type tcglogPhase int
+
+const (
+	// tcglogPhasePreOSBeforeMeasureSecureBootConfig is the pre-OS phase of the log
+	// before measurement of the secure boot configuration has begun. This is the
+	// starting point.
+	tcglogPhasePreOSBeforeMeasureSecureBootConfig tcglogPhase = iota
+
+	// tcglogPhasePreOSMeasuringSecureBootConfig is the pre-OS phase of the log where
+	// the secure boot configuration is being measured.
+	// tcglogPhasePreOSBeforeMeasureSecureBootConfig transitions to this phase by the
+	// first non EV_SEPARATOR event in PCR7.
+	tcglogPhasePreOSMeasuringSecureBootConfig tcglogPhase = iota
+
+	// tcglogPhasePreOSAfterMeasureSecureBootConfig is the pre-OS phase of the log after
+	// measuring the secure boot configuration and which may contain authentication and
+	// loading of pre-OS components. An EV_SEPARATOR in PCR7 transitions
+	// tcglogPhasePreOSMeasuringSecureBootConfig to this phase. Some older firmware
+	// implementations skip this phase because they measure EV_SEPARATORS in PCRs 0-7 to
+	// indicate the transition to OS-present.
+	tcglogPhasePreOSAfterMeasureSecureBootConfig
+
+	// tcglogPhaseTransitioningToOSPresent describes the phase of the log where the transition
+	// to OS-present happens. Either of the 2 previous pre-OS phases can transition to this one
+	// by an EV_SEPARATOR event in PCRs 0-6.
+	tcglogPhaseTransitioningToOSPresent
+
+	// tcglogPhaseOSPresent describes the phase of the log where the OS is in control. The
+	// tcglogPhaseTransitioningToOSPresent phase transitions to this when all EV_SEPARATOR
+	// events in PCRs 0-7 have been seen.
+	tcglogPhaseOSPresent
+
+	// tcglogPhaseErr indicates that a previous event was interpreted as an error. Attempting
+	// to process the rest of the log will cause a panic.
+	tcglogPhaseErr
+)
+
+// tcglogPhaseTracker tracks the phases of a TCG log up to OS-present, and processes
+// events to validate the general consistency of the structure of the log and sequence
+// of events.
+type tcglogPhaseTracker struct {
+	phase             tcglogPhase
+	numSeparatorsSeen int
+	pcrSeparatorsSeen map[tpm2.Handle]struct{}
+}
+
+// newTcgLogPhaseTracker returns a new tcglogPhaseTracker
+func newTcgLogPhaseTracker() *tcglogPhaseTracker {
+	return &tcglogPhaseTracker{
+		phase:             tcglogPhasePreOSBeforeMeasureSecureBootConfig,
+		numSeparatorsSeen: 0,
+		pcrSeparatorsSeen: make(map[tpm2.Handle]struct{}),
```

Comment:
nitpick: we tend to just use bool and not empty struct for this kind of set structures

Reply:
I can change that.